Library/OcCpuLib/OcCpuLib.c

```diff
@@ -1418,35 +1417,39 @@ OcCpuScanProcessor (
   //  1. CPUID usage for interaction between Hypervisors and Linux.: https://lwn.net/Articles/301888/
   //  2. [Qemu-devel] [PATCH v2 0/3] x86-kvm: Fix Mac guest timekeeping by exposi: https://lists.gnu.org/archive/html/qemu-devel/2017-01/msg04344.html
   //
-  if (Cpu->Hypervisor && (Cpu->Features & CPUID_EXTFEATURE_TSCI)){
+  AsmCpuid (0x40000000, &CpuidEax, NULL, NULL, NULL);
+  if (Cpu->Hypervisor && (CpuidEax >= 0x40000010)){
     DEBUG ((DEBUG_INFO, "OCCPU: Cpu under virtualization, try get TSC/FSB frequency from VMWare Timing\n"));
     AsmCpuid (0x40000010, &CpuidEax, &CpuidEbx, NULL, NULL);
-    //
-    // We get kHZ from node and we should translate it first.
-    //
-    Cpu->CPUFrequencyFromTSC = CpuidEax * 1000;
-    Cpu->FSBFrequency = CpuidEbx * 1000;
 
-    Cpu->CPUFrequency = Cpu->CPUFrequencyFromTSC;
-    //
-    // We can caculate Bus Ratio here
-    //
-    Cpu->MaxBusRatio = DivU64x32 (Cpu->CPUFrequency, Cpu->FSBFrequency);
-    //
-    // We don't have anything like turbo, so we just assign some variables here
-    //
-    Cpu->MinBusRatio = Cpu->MaxBusRatio;
-    Cpu->CurBusRatio = Cpu->CurBusRatio;
+    if (CpuidEax & CpuidEbx){
```

Comment:
Should we get fallback assignment or just go to normal procedure like before if this failed(like hypervisor didn't expose VMWare Timing Node)?
If we go to normal procedure, we can get CPU Frequency but definately the MaxRatio will be wrong.
The Intel part will work but the ryzen part will failed like before.

Reply:
I would say fallback to native coding, as it seems to me a hypervisor bug to support neither.

Reply:
Also, please add a space between ) and {, like in `if (Cpu->Hypervisor && (CpuidEax >= 0x40000010)){`.